crates/polars-ops/src/series/ops/interpolation/interpolate_by.rs

```diff
@@ -0,0 +1,327 @@
+
+use super::linear_itp;
+
+#[inline]
+fn signed_interp_by_sorted<T, F>(y_start: T, y_end: T, x: &[F], out: &mut Vec<T>)
+where
+    T: Sub<Output = T>
+        + Mul<Output = T>
+        + Add<Output = T>
+        + Div<Output = T>
+        + NumCast
+        + Copy
+        + Zero,
+    F: Sub<Output = F> + NumCast + Copy,
+{
+    let range_y = y_end - y_start;
+    let range_x = NumCast::from(x[x.len() - 1] - x[0]).unwrap();
+    let slope = range_y / range_x;
+    let x_start = x[0];
+    for x_i in &x[1..x.len() - 1] {
+        let x_delta = NumCast::from(*x_i - x_start).unwrap();
+        let v = linear_itp(y_start, x_delta, slope);
+        out.push(v)
+    }
+}
+
+#[inline]
+fn signed_interp_by<T, F>(
+    y_start: T,
+    y_end: T,
+    x: &[F],
+    out: &mut [T],
+    sorting_indices: &[IdxSize],
+    low_idx: usize,
+) where
+    T: Sub<Output = T>
+        + Mul<Output = T>
+        + Add<Output = T>
+        + Div<Output = T>
+        + NumCast
+        + Copy
+        + Zero,
+    F: Sub<Output = F> + NumCast + Copy,
+{
+    let range_y = y_end - y_start;
+    let range_x = NumCast::from(x[x.len() - 1] - x[0]).unwrap();
+    let slope = range_y / range_x;
+    let x_start = x[0];
+    for (offset, x_i) in (x[1..x.len() - 1]).iter().enumerate() {
+        let x_delta = NumCast::from(*x_i - x_start).unwrap();
+        let v = linear_itp(y_start, x_delta, slope);
+        let out_idx = unsafe { sorting_indices.get_unchecked(low_idx + offset + 1) };
+        unsafe { *out.get_unchecked_mut(*out_idx as usize) = v };
+    }
+}
+
+fn interpolate_impl_by_sorted<T, F, I>(
+    chunked_arr: &ChunkedArray<T>,
+    by: &ChunkedArray<F>,
+    interpolation_branch: I,
+) -> PolarsResult<ChunkedArray<T>>
+where
+    T: PolarsNumericType,
+    F: PolarsIntegerType,
+    I: Fn(T::Native, T::Native, &[F::Native], &mut Vec<T::Native>),
+{
+    // This implementation differs from pandas as that boundary None's are not removed.
+    // This prevents a lot of errors due to expressions leading to different lengths.
+    if !chunked_arr.has_validity() || chunked_arr.null_count() == chunked_arr.len() {
+        return Ok(chunked_arr.clone());
+    }
+
+    // At the moment, the algorithm assumes no missing values in the `by` column.
+    let by = by.rechunk();
+    let by_values = by.cont_slice().map_err(|_| {
+        polars_err!(
+            InvalidOperation:
+            "null values in `by` column are not yet supported in 'interpolate_by' expression"
+        )
+    })?;
+
+    // We first find the first and last so that we can set the null buffer.
+    let first = chunked_arr.first_non_null().unwrap();
+    let last = chunked_arr.last_non_null().unwrap() + 1;
+
+    // Fill out with first.
+    let mut out = Vec::with_capacity(chunked_arr.len());
+    let mut iter = chunked_arr.into_iter().enumerate().skip(first);
+    for _ in 0..first {
+        out.push(Zero::zero())
+    }
+
+    let mut low_val = None;
+    let mut low_idx = None;
```

Comment:
I believe we can set `low_idx` on 0. That would save an `unwrap` branch in hot loop.

Not sure, but could `low_val` be set to a first value as well?